### ISSUE TITLE
jepsen "batch set crash" failed with 1.0.2 RC

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#1222]
## Installation details

Scylla version (or git commit hash): 60307f62fe82ef622dc0623d02673dc5de3afd97 
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI): Debian jessie
## Hardware details (for performance issues)

Platform (physical/VM/cloud instance type/docker): Amazon c3.8xlarge instance
## Description

Jepsen's `batch set crash` creates the keyspace and tables as described below and runs the batch transactions while periodically crashing (`kill -9` nodes):

```
jepsen_keyspace: SimpleStrategy with RF 3
       (cql/create-table conn "a"
                          (column-definitions {:id :int
                                               :value :int
                                               :primary-key [:id]}))
        (cql/create-table conn "b"
                          (column-definitions {:id :int
                                               :value :int
                                               :primary-key [:id]}))
```

Insertion is performed in a following way:

```
 (with-consistency-level ConsistencyLevel/QUORUM
                    (cql/atomic-batch conn (queries
                                            (insert-query "a" {:id value
                                                               :value value})
                                            (insert-query "b" {:id (- value)
                                                               :value value})))))
```

At the end of the test the final reading is performed as follow after a 65 seconds delay:

```
:read (try (let [value-a (->> (with-retry-policy aggressive-read
                                      (with-consistency-level ConsistencyLevel/ALL
                                        (cql/select conn "a")))
                                    (map :value)
                                    (into (sorted-set)))
                       value-b (->> (with-retry-policy aggressive-read
                                      (with-consistency-level ConsistencyLevel/ALL
                                        (cql/select conn "b")))
                                    (map :value)
                                    (into (sorted-set)))]
                   (if-not (= value-a value-b)
                     (assoc op :type :fail :value [value-a value-b])
                     (assoc op :type :ok :value value-a)))
                 (catch UnavailableException e
                   (info "Not enough replicas - failing")
                   (assoc op :type :fail :value (.getMessage e)))
                 (catch ReadTimeoutException e
                   (assoc op :type :fail :value :timed-out))
                 (catch NoHostAvailableException e
                   (info "All nodes are down - sleeping 2s")
                   (Thread/sleep 2000)
                   (assoc op :type :fail :value (.getMessage e))))))
```

which effectively reads all values from both table `a` and table `b` and compares these two sets which are expected to be equal since batchlog replayer is expected to replay all pending mutations after a 65s the test gives it at the end after all the connectivity is restored.

As a result there is a value (714) that is missing in a table `a`.
## Log files

[history.txt](https://github.com/scylladb/scylla/files/240122/history.txt)
[n1-system-log.txt](https://github.com/scylladb/scylla/files/240121/n1-system-log.txt)
[n2-system-log.txt](https://github.com/scylladb/scylla/files/240126/n2-system-log.txt)
[n3-system-log.txt](https://github.com/scylladb/scylla/files/240123/n3-system-log.txt)
[n4-system-log.txt](https://github.com/scylladb/scylla/files/240125/n4-system-log.txt)
[n5-system-log.txt](https://github.com/scylladb/scylla/files/240124/n5-system-log.txt)
[results.txt](https://github.com/scylladb/scylla/files/240127/results.txt)

**Comment by gleb-cloudius** [comment]
How often the test kills nodes and how many simultaneously.  In n5-system-log.txt I see up to three node dead simultaneously, I do not think batchlog will work in that scenario especially as it is not saved to persistent storage for 10 seconds:

INFO  2016-04-27 13:36:59,829 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,175 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,175 [shard 3] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,175 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,175 [shard 2] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 2] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 2] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 3] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 3] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 3] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 1] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 1] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 0] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 1] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,176 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:00,830 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:01,830 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:02,176 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:02,176 [shard 0] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:02,830 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:02,831 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 2] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 3] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 2] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 3] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 1] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 1] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 0] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,177 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:03,831 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:04,831 [shard 0] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:04,831 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 2] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 2] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 3] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 3] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 1] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,177 [shard 1] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,178 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:05,832 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:06,832 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:07,178 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:07,178 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:07,832 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:07,832 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 0] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 2] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 1] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 3] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 2] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 1] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,179 [shard 3] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,833 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:08,833 [shard 0] rpc - client 192.168.122.12: fail to connect: Connection refused
INFO  2016-04-27 13:37:09,833 [shard 0] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,179 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,179 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 0] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 0] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 2] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 3] rpc - client 192.168.122.13: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 3] rpc - client 192.168.122.14: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 1] rpc - client 192.168.122.11: fail to connect: Connection refused
INFO  2016-04-27 13:37:10,180 [shard 1] rpc - client 192.168.122.14: fail to connect: Connection refused

**Comment by vladzcloudius** [comment]
On 05/04/2016 03:44 PM, Gleb Natapov wrote:
>
> How often the test kills nodes and how many simultaneously. In 
> n5-system-log.txt I see up to three node dead simultaneously,
>

The test chooses the random subset of all nodes for killing. See 
/history.txt/ for a specific choices in this particular run. U may 
notice that 3 nodes is not a limit - it kills all 5 in the first nemesis 
round. ;)


> I do not think batchlog will work in that scenario especially as it is 
> not saved to persistent storage for 10 seconds:
>

Hmmm... According to DataStacks BATCH documentation if one of batched 
operations succeeds - all will succeed:

Batches are logged by default. In the context of a Cassandra batch operation, logged
batches means that ifany of the batch succeeds, all of it will 
<http://www.datastax.com/dev/blog/atomic-batches-in-cassandra-1-2>. To achieve this,
Cassandra first writes the serialized batch to the batchlog system table that consumes the
serialized batch as blob data.


AFAIU in order to stand behind this contract the write to the system 
table above has to either flushed to persistent storage or be written to 
a commitlog *before* any of the batched operations are invoked.
Otherwise the above contract may be simply broken by the /kill/-scenario 
and frankly I don't understand why did u care about how many nodes were 
killed simultaneously - it's enough to kill a single node - the one that 
is a batch coordinator.
So if what u describe "(at least one of) the batched operation was 
issued before the batch logging was committed" actually takes place the 
this clearly breaks the contract above.

**Comment by vladzcloudius** [comment]
@gleb-cloudius I can still reproduce the issue with `--commitlog-sync batch --commitlog-sync-batch-window-in-ms 1`.

**Comment by elcallio** [comment]
So, I've read through the issue and the files, and the discussion on the 
mailing list. Not sure what the considered approach here is? Is the 
working theory that this is due to lost commitlog data, or a more high 
level batchlog bug?
I can work on fixing the batch mode for commit log to work more as Avi 
suggested...

Not sure if trying to repro this with jepsen helps very much either, 
since ti feels like any errors quickly get lost in the sheer volume...

On a side note btw, are there any plans for an Israel conference this 
year as well? If there are, it would be neat to know, what with 
vacations etc coming up.

/C

**Comment by avikivity** [comment]
So far we had two suggestions:

 1. add an in-memory log of in-flight buffers, a monitoring process (using fork + MAP_SHARED, or perhaps /dev/hugepages), on signal replay the log
 2. fix up batch mode and introduce an intermediate mode where we ack after IO_CMD_PWRITE completes (but before fsync)

Option 2 is much safer (in the development risk sense) in that almost all of the work is needed for real batch mode (ack after fsync).  Option 1 is somewhat baroque.  So I think we should start with that.

wrt. an internal conference, please ask on the internal list

**Comment by gleb-cloudius** [comment]
On Mon, May 16, 2016 at 03:46:25AM -0700, Calle Wilund wrote:
> So, I've read through the issue and the files, and the discussion on the 
> mailing list. Not sure what the considered approach here is? Is the 
> working theory that this is due to lost commitlog data, or a more high 
> level batchlog bug?
The test as it is cannot work because of commitlog data lose, but this
does not mean there is no other bugs that hide behind the obvious one.
Vlad tried to use commitlog batch mode where it should be synced before write
is marked as done, but the test still failed. It may indicate a bug in
the batch mode since it is not test very well (if at all).

--
			Gleb.

**Comment by elcallio** [comment]
Are we 100% sure it is commit log fails here. I'm looking at the storage_proxy code, and one glaring thing about it is that it does

1.) write batch log entry
2.) begin the batched actual mutations
3.) issue batch log entry removal

However, it does, afaict, _not_ wait for 2) to finish before 3) (which is what origin does). So we remove the batch log before we've verified the ops went through... 

I'd love to have a more convenient test here. :-)

**Comment by gleb-cloudius** [comment]
On Mon, May 16, 2016 at 04:04:58AM -0700, Calle Wilund wrote:
> Are we 100% sure it is commit log fails here. I'm looking at the storage_proxy code, and one glaring thing about it is that it does
> 
> 1.) write batch log entry
> 2.) begin the batched actual mutations
> 3.) issue batch log entry removal
> 
> However, it does, afaict, _not_ wait for 2) to finish before 3) (which is what origin does). So we remove the batch log before we've verified the ops went through... 
> 
As far as I see it does wait. What makes you think otherwise?

--
			Gleb.

**Comment by avikivity** [comment]
Are you referring to mutate_begin()?

I think it does wait:

```
future<> storage_proxy::mutate_begin(std::vector<unique_response_handler> ids, db::consistency_level cl) {
    return parallel_for_each(ids, [this, cl] (unique_response_handler& protected_response) {
        auto response_id = protected_response.id;
        // it is better to send first and hint afterwards to reduce latency
        // but request may complete before hint_to_dead_endpoints() is called and
        // response_id handler will be removed, so we will have to do hint with separate
        // frozen_mutation copy, or manage handler live time differently.
        hint_to_dead_endpoints(response_id, cl);

        auto timeout = clock_type::now() + std::chrono::milliseconds(_db.local().get_config().write_request_timeout_in_ms());
        // call before send_to_live_endpoints() for the same reason as above
        auto f = response_wait(response_id, timeout);
        send_to_live_endpoints(protected_response.release(), timeout); // response is now running and it will either complete or timeout
        return std::move(f);
    });
}
```

`f` waits for the response.  The function may be badly named.

**Comment by gleb-cloudius** [comment]
On Mon, May 16, 2016 at 02:08:32PM +0300, Gleb Natapov wrote:
> On Mon, May 16, 2016 at 04:04:58AM -0700, Calle Wilund wrote:
> > Are we 100% sure it is commit log fails here. I'm looking at the storage_proxy code, and one glaring thing about it is that it does
> > 
> > 1.) write batch log entry
> > 2.) begin the batched actual mutations
> > 3.) issue batch log entry removal
> > 
> > However, it does, afaict, _not_ wait for 2) to finish before 3) (which is what origin does). So we remove the batch log before we've verified the ops went through... 
> > 
> As far as I see it does wait. What makes you think otherwise?
> 
Hmm, may be because of mutate_begin() name, but the function return
ready future only after CL is reached for the whole batch.

--
			Gleb.

**Comment by elcallio** [comment]
Ah, fair enough. My bad. Meh.

**Comment by tzach** [comment]
@elcallio any update on this issue?

**Comment by elcallio** [comment]
So, this took a while, yet spent most of the time doing nothing. (I.e. looking at test runs and aimlessly trying to figure out what we/origin is doing). 

I created a dtest that does the same thing as the jepsen test mentioned above and ran it (repeatedly, listlessly and life-suckingly) with scylla/origin. I also added some extra checks in the test, thus it basically does this:

- Start three nodes with commitlog mode BATCH (must have)
- Create two identical CF:s (a and b)in the same keyspace
- Repeatedly insert the same value into the same partition for a&b using a logged batch
- In the background, randomly kill some of the nodes
- Continue into actual batch execution has failed X times
- Restart all nodes
- Query the contents of a&b and verify their contents are the same (with CL=ALL)
- Query the contents of a&b and verify their row count is equal to number of insertions (with CL=ALL)

Now, origin manages this. However, if one replaces the final query CL=ALL with CL=ONE (default), it failes criteria two. I.e. row contents is consistent per node, but not across the cluster. (Reasonable). 
Scylla however will here give back inconsistent data per node. I.e. a & b have different contents in a single node. 

Investigating and haxxor-fixing things came up with a few conclusions:

1. Batch log replay uses the wrong consistency level. It should replay on something akin to CL=ALL (and actually even harsher, it should require all live target nodes to respond - in practice ALL, but not a 100% contract I think). 
2. Scylla commit log and replay is **not** at fault. In fact, at least disabling the short-cuttting (skipping) of mutations in replay, all writes put in and flushed (thus acked) are also replayed on node restart. There could still be some issue with our RP-storing for flushed sstables...
However, our batch mode is neither very fast nor very batch, since it writes a single mutation on each request. This logic flow should at least  be better serialized (right now I've been running with a modded variant that explicitly makes sure we never accidentally reorder a flush). Me thinks a write queue. 

**The interesting part:**

The problem turns out to be exactly the disparity in batch statement that Pawel pointed out a few days ago: Origin will, for a batch statement, per key space, merge all mutation that reference the same PK into a single mutation (origins mutations can affect more than one CF). This means that for the above case, doing an insert with the same PK and value for CF a&b will in fact be _a single mutation_ in origin. Whereas for us, it will in fact be two, and thus two parallel requests in storage proxy. 
So, what can happen (and does) is that the batch query (which is executed at CL=quorum), results in two requests, each going to three nodes, and waiting for two answers. A node can thus happen to process and ack one of the two, then get killed before the second is processed. However, since other nodes might answer, the coordinator will still get enough ACK:s to consider CL fulfilled. Thus the batch is considered successful, and will also be removed from log. But individual nodes might still miss data points for one of the CF:s, which in turn is why we get inconsistent data upon query with CL=ONE. 

Indeed, modifying origin to send one mutation per insert, similar to scylla, shows the exact same behaviour. It also means that had the batch been in two key spaces, we'd get the same inconsistent data. 

My guess as to why the original test (which was afaik run on 5 nodes with RF=3) fails even on final query with CL=ALL is a combo of the above and issue (1). 

I am however not sure how to deal with the "mutation merging" disparity. Our model does not really support it. The best we could do is perhaps allow more than one mutation per RPC message. 
Otoh, why batch statements should be more consistent per keyspace than across them is not immediately obvious to me either. 
In any case, I can at least make patches for 1&2 above.

**Comment by vladzcloudius** [comment]
On 06/07/2016 05:05 PM, Calle Wilund wrote:
>
> So, this took a while, yet spent most of the time doing nothing. (I.e. 
> looking at test runs and aimlessly trying to figure out what we/origin 
> is doing).
>
> I created a dtest that does the same thing as the jepsen test 
> mentioned above and ran it (repeatedly, listlessly and life-suckingly) 
> with scylla/origin. I also added some extra checks in the test, thus 
> it basically does this:
>
>   * Start three nodes with commitlog mode BATCH (must have)
>   * Create two identical CF:s (a and b)in the same keyspace
>   * Repeatedly insert the same value into the same partition for a&b
>     using a logged batch
>   * In the background, randomly kill some of the nodes
>   * Continue into actual batch execution has failed X times
>   * Restart all nodes
>   * Query the contents of a&b and verify their contents are the same
>     (with CL=ALL)
>   * Query the contents of a&b and verify their row count is equal to
>     number of insertions (with CL=ALL)
>
> Now, origin manages this. However, if one replaces the final query 
> CL=ALL with CL=ONE (default), it failes criteria two. I.e. row 
> contents is consistent per node, but not across the cluster. 
> (Reasonable).
> Scylla however will here give back inconsistent data per node. I.e. a 
> & b have different contents in a single node.
>
> Investigating and haxxor-fixing things came up with a few conclusions:
>
>  1. Batch log replay uses the wrong consistency level. It should
>     replay on something akin to CL=ALL (and actually even harsher, it
>     should require all live target nodes to respond - in practice ALL,
>     but not a 100% contract I think).
>  2. Scylla commit log and replay is *not* at fault. In fact, at least
>     disabling the short-cuttting (skipping) of mutations in replay,
>     all writes put in and flushed (thus acked) are also replayed on
>     node restart. There could still be some issue with our RP-storing
>     for flushed sstables... However, our batch mode is neither very
>     fast nor very batch, since it writes a single mutation on each
>     request. This logic flow should at least be better serialized
>     (right now I've been running with a modded variant that explicitly
>     makes sure we never accidentally reorder a flush). Me thinks a
>     write queue.
>
> *The interesting part:*
>
> The problem turns out to be exactly the disparity in batch statement 
> that Pawel pointed out a few days ago: Origin will, for a batch 
> statement, per key space, merge all mutation that reference the same 
> PK into a single mutation (origins mutations can affect more than one 
> CF). This means that for the above case, doing an insert with the same 
> PK and value for CF a&b will in fact be /a single mutation/ in origin. 
> Whereas for us, it will in fact be two, and thus two parallel requests 
> in storage proxy.
> So, what can happen (and does) is that the batch query (which is 
> executed at CL=quorum), results in two requests, each going to three 
> nodes, and waiting for two answers. A node can thus happen to process 
> and ack one of the two, then get killed before the second is 
> processed. However, since other nodes might answer, the coordinator 
> will still get enough ACK:s to consider CL fulfilled. Thus the batch 
> is considered successful, and will also be removed from log. But 
> individual nodes might still miss data points for one of the CF:s, 
> which in turn is why we get inconsistent data upon query with CL=ONE.
>
> Indeed, modifying origin to send one mutation per insert, similar to 
> scylla, shows the exact same behaviour. It also means that had the 
> batch been in two key spaces, we'd get the same inconsistent data.
>
> My guess as to why the original test (which was afaik run on 5 nodes 
> with RF=3) fails even on final query with CL=ALL is a combo of the 
> above and issue (1).
>

Could u, pls., clarify what combo do u refer here? AFAIKT , if QUORUM CL 
is reached during a replay a final read with CL=ALL should not see any 
inconsistent data.

**Comment by elcallio** [comment]
Right now batch replay is done with CL=ANY. Also worth remembering is that even if a replay is done, it will suffer from the same issues as the original batch operation, i.e. the two inserts will be two different mutations, subject to the same potentially disparate responses/CL acknowledgement.

**Comment by vladzcloudius** [comment]
On 06/07/2016 05:37 PM, Calle Wilund wrote:
>
> Right now batch replay is done with CL=ANY. Also worth remembering is 
> that even if a replay is done, it will suffer from the same issues as 
> the original batch operation, i.e. the two inserts will be two 
> different mutations, subject to the same potentially disparate 
> responses/CL acknowledgement.
>

In the previous (long) email u wrote "So, what can happen (and does) is 
that the batch query (which is executed at CL=quorum)". Now u say it's 
executed with CL=ANY. I'm confused.

**Comment by gleb-cloudius** [comment]
On Tue, Jun 07, 2016 at 07:05:32AM -0700, Calle Wilund wrote:
> 
> 1. Batch log replay uses the wrong consistency level. It should replay on something akin to CL=ALL (and actually even harsher, it should require all live target nodes to respond - in practice ALL, but not a 100% contract I think). 
I remember wondering about this ANY already:
https://groups.google.com/forum/#!msg/scylladb-dev/dbkT0TgQuF8/-LuvTvqeBgAJ

In what way is it different from CL=ALL?

> 
> I am however not sure how to deal with the "mutation merging" disparity. Our model does not really support it.
Why our model does not support it? Why we cannot merge mutations on a
coordinator?

--
			Gleb.

**Comment by gleb-cloudius** [comment]
On Tue, Jun 07, 2016 at 08:10:23AM -0700, vladzcloudius wrote:
> 
> 
> On 06/07/2016 05:37 PM, Calle Wilund wrote:
> >
> > Right now batch replay is done with CL=ANY. Also worth remembering is 
> > that even if a replay is done, it will suffer from the same issues as 
> > the original batch operation, i.e. the two inserts will be two 
> > different mutations, subject to the same potentially disparate 
> > responses/CL acknowledgement.
> >
> 
> In the previous (long) email u wrote "So, what can happen (and does) is 
> that the batch query (which is executed at CL=quorum)". Now u say it's 
> executed with CL=ANY. I'm confused.
> 
CL=ANY is for replay, not original query which is done with QUORUM. 

--
			Gleb.

**Comment by vladzcloudius** [comment]
On 06/07/2016 06:17 PM, Gleb Natapov wrote:
> On Tue, Jun 07, 2016 at 08:10:23AM -0700, vladzcloudius wrote:
> >
> >
> > On 06/07/2016 05:37 PM, Calle Wilund wrote:
> > >
> > > Right now batch replay is done with CL=ANY. Also worth remembering is
> > > that even if a replay is done, it will suffer from the same issues as
> > > the original batch operation, i.e. the two inserts will be two
> > > different mutations, subject to the same potentially disparate
> > > responses/CL acknowledgement.
> > >
> >
> > In the previous (long) email u wrote "So, what can happen (and does) is
> > that the batch query (which is executed at CL=quorum)". Now u say it's
> > executed with CL=ANY. I'm confused.
> >
> CL=ANY is for replay, not original query which is done with QUORUM.

Oh, I see. Thanks for a clarification, Gleb.

**Comment by elcallio** [comment]
Gleb, CL=ANY means we wait for "target KS replication factor" number of nodes, where nodes are determined by repl. the strategy "get_natural_endpoints" return value. Origin batch replay request handler otoh explicitly waits for all reachable natural + pending endpoints, and writes hints to those that are down. In most cases I assume the difference is none, but I'd still say that it is not guaranteed to be (at least not with a strategy != simple/net). 

As for mutations; The problem is that our mutation only allows modification of a single CF. Origin otoh allows multiple CF:s as long as they share a keyspace. Since origin does a merge into a single mutation of all batch statement sub-statements using the same PK value, in effect an insert that is "symmetrical" such as that in the test, will be both sent using a single message, and also applied to the commit log as a single, "atomic" operation.

**Comment by gleb-cloudius** [comment]
On Tue, Jun 07, 2016 at 08:43:30AM -0700, Calle Wilund wrote:
> Gleb, CL=ANY means we wait for "target KS replication factor" number of nodes, where nodes are determined by repl. the strategy "get_natural_endpoints" return value.
I assume you mean CL=ALL here.

> Origin batch replay request handler otoh explicitly waits for all reachable natural + pending endpoints, and writes hints to those that are down. In most cases I assume the difference is none, but I'd still say that it is not guaranteed to be (at least not with a strategy != simple/net). 
If I am reading origin correctly they do wait for all replicas (see
ReplayWriteResponseHandler::totalBlockFor()), but the funny thing is that
if there are dead replicas they will not send to all of them, so they
are doomed to fail.

> 
> As for mutations; The problem is that our mutation only allows modification of a single CF. Origin otoh allows multiple CF:s as long as they share a keyspace. Since origin does a merge into a single mutation of all batch statement sub-statements using the same PK value, in effect an insert that is "symmetrical" such as that in the test, will be both sent using a single message, and also applied to the commit log as a single, "atomic" operation. 
> 
What if PK value is the same but modification is different? How can they
be merged?

--
			Gleb.

**Comment by gleb-cloudius** [comment]
On Tue, Jun 07, 2016 at 07:05:32AM -0700, Calle Wilund wrote:
> My guess as to why the original test (which was afaik run on 5 nodes with RF=3) fails even on final query with CL=ALL is a combo of the above and issue (1). 
> 
I do not see a scenario where combination of those two can fail CL=ALL
read. Can you elaborate?

--
			Gleb.

**Comment by elcallio** [comment]
On 2016-06-07 16:50, Gleb Natapov wrote:
> On Tue, Jun 07, 2016 at 08:43:30AM -0700, Calle Wilund wrote:
> > Gleb, CL=ANY means we wait for "target KS replication factor" number 
> of nodes, where nodes are determined by repl. the strategy 
> "get_natural_endpoints" return value.
> I assume you mean CL=ALL here.
>
Indeed.
> > Origin batch replay request handler otoh explicitly waits for all 
> reachable natural + pending endpoints, and writes hints to those that 
> are down. In most cases I assume the difference is none, but I'd still 
> say that it is not guaranteed to be (at least not with a strategy != 
> simple/net).
> If I am reading origin correctly they do wait for all replicas (see
> ReplayWriteResponseHandler::totalBlockFor()), but the funny thing is that
> if there are dead replicas they will not send to all of them, so they
> are doomed to fail.
>
They send and expect answer from all target nodes marked as up, but weed 
out those marked down and send hints to them instead, which I guess is 
more or less what the normal send should do as well (except we don't 
support hints yet).
Bottom line is that the batch replay handler is sort of a 
strategy-independent cl-all.
> >
> > As for mutations; The problem is that our mutation only allows 
> modification of a single CF. Origin otoh allows multiple CF:s as long 
> as they share a keyspace. Since origin does a merge into a single 
> mutation of all batch statement sub-statements using the same PK 
> value, in effect an insert that is "symmetrical" such as that in the 
> test, will be both sent using a single message, and also applied to 
> the commit log as a single, "atomic" operation.
> >
> What if PK value is the same but modification is different? How can they
> be merged?
>
Origin "Mutation" is just a collection of modifications to CF:s within a 
single key space. Our lingo is different from origins, but it is 
basically a map CF-id -> Partition. The partition in turn contain cells 
which makes up the mutation, i.e. data or tombstones. As long as the 
modifications (cells) are within the same PK, it can contain any number 
of changes in a single mutation. Otoh, a mutation in origin is limited 
to a single PK type+value.

**Comment by elcallio** [comment]
On 2016-06-07 16:52, Gleb Natapov wrote:
> On Tue, Jun 07, 2016 at 07:05:32AM -0700, Calle Wilund wrote:
> > My guess as to why the original test (which was afaik run on 5 nodes 
> with RF=3) fails even on final query with CL=ALL is a combo of the 
> above and issue (1).
> >
> I do not see a scenario where combination of those two can fail CL=ALL
> read. Can you elaborate?
>
To be perfectly honest, neither can I. It might be that the end issue of 
getting broken results with CL=ALL is unrelated.
The discrepancy I've found can cause us to have unsymmetrical views of 
two CF:s across replicas. This should however, assuming that we do in 
fact query all replica nodes in the final selects, not prevent a proper 
merge of all the data.

**Comment by gleb-cloudius** [comment]
On Wed, Jun 08, 2016 at 01:04:08AM -0700, Calle Wilund wrote:
> 
> 
> On 2016-06-07 16:50, Gleb Natapov wrote:
> > On Tue, Jun 07, 2016 at 08:43:30AM -0700, Calle Wilund wrote:
> > > Gleb, CL=ANY means we wait for "target KS replication factor" number 
> > of nodes, where nodes are determined by repl. the strategy 
> > "get_natural_endpoints" return value.
> > I assume you mean CL=ALL here.
> >
> Indeed.
> > > Origin batch replay request handler otoh explicitly waits for all 
> > reachable natural + pending endpoints, and writes hints to those that 
> > are down. In most cases I assume the difference is none, but I'd still 
> > say that it is not guaranteed to be (at least not with a strategy != 
> > simple/net).
> > If I am reading origin correctly they do wait for all replicas (see
> > ReplayWriteResponseHandler::totalBlockFor()), but the funny thing is that
> > if there are dead replicas they will not send to all of them, so they
> > are doomed to fail.
> >
> They send and expect answer from all target nodes marked as up, but weed 
Right, I misread

            protected int totalBlockFor()
            {
                return this.naturalEndpoints.size();
            }

as checking for all natural endpoints, but hear natural endpoints
contain only live endpoins.

> out those marked down and send hints to them instead, which I guess is 
> more or less what the normal send should do as well (except we don't 
> support hints yet).
Normal send will not write hints unless there is enough nodes to satisfy
CL.

> Bottom line is that the batch replay handler is sort of a 
> strategy-independent cl-all.
cl-all is always strategy independent, you write to all and expect reply
from all.

What they do is write with CL=ONE not ANY or ALL.

--
			Gleb.

**Comment by tzach** [comment]
@elcallio any updates on this issue?

**Comment by elcallio** [comment]
A number of patches where committed, for issues related to batch 
consistency. The dtest variant of the test I created works fine. Still 
waiting for results on this particular (jepsen) test though.


Den 2016-07-10 kl. 10:38, skrev Tzach Livyatan:

**Comment by tzach** [comment]
@haarcuba can you reproduce with latest version?

**Comment by haarcuba** [comment]
I have a failing  'cassandra batch set crash bootstrap' a.k.a batch-crash-subset-bootstrap
I think it's the same one. @vladzcloudius ?

**Comment by vladzcloudius** [comment]
@haarcuba The test the issue is about was not doing bootstrap. The one that u've refered does the same things but IN ADDITION performs bootstrap cycle in parallel. 

We had separate issues with bootstrapping those days therefore i intentionally used this test in order not to hit know bootstrap issues.

**Comment by haarcuba** [comment]
OK. I'll see if I can reproduce the original failure

**Comment by haarcuba** [comment]
I managed to reproduce it after 20+ trials.
[batch-crash-subset.fail.log.txt](https://github.com/scylladb/scylla/files/453873/batch-crash-subset.fail.log.txt)

**Comment by kostja** [comment]
Irrelevant, we rewrote Jepsen in 2020
